sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,388 @@
+  , secretKeyFromByteString
+  , secretKeyToBinary
+  , secretKeyToHexByteString
+  , secretKeyToHexText
+
+    -- ** Nonce
+  , Nonce
+  , nonceFromByteString
+  , nonceToBinary
+  , nonceToHexByteString
+  , nonceToHexText
+
+    -- ** Hash
+  , Hash
+  , hashFromByteString
+  , hashToBinary
+  , hashToHexByteString
+  , hashToHexText
+
+    -- ** Encryption and Decryption
+  , encrypt
+  , decrypt
+  ) where
+
+import Control.Monad (void)
+import Data.ByteString (StrictByteString)
+import qualified Data.ByteString.Base16 as Base16
+import qualified Data.ByteString.Internal as BS
+import qualified Data.ByteString.Unsafe as BS
+import Data.Text (Text)
+import Data.Text.Display (Display, displayBuilder)
+import qualified Data.Text.Lazy.Builder as Builder
+import Data.Word (Word8)
+import Foreign (ForeignPtr)
+import qualified Foreign
+import Foreign.C (CChar, CSize, CUChar, CULLong)
+import GHC.IO.Handle.Text (memcpy)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+import LibSodium.Bindings.Random (randombytesBuf)
+import LibSodium.Bindings.Secretbox (cryptoSecretboxEasy, cryptoSecretboxKeyBytes, cryptoSecretboxKeygen, cryptoSecretboxMACBytes, cryptoSecretboxNonceBytes, cryptoSecretboxOpenEasy)
+import Sel.Internal
+
+-- $introduction
+-- "Authenticated Encryption" uses a secret key along with a single-use number
+-- called a "nonce" to encrypt a message.
+-- The resulting hash is accompanied by an authentication tag.
+--
+-- Encryption is done with the XSalsa20 stream cipher and authentication is done with the
+-- Poly1305 MAC hash.
+
+-- $usage
+--
+-- > import qualified Sel.SecretKey.AuthenticatedEncryption as AuthenticatedEncryption
+-- >
+-- > main = do
+-- >   -- We get the secretKey from the other party or with 'newSecretKey'.
+-- >   -- We get the nonce from the other party with the message, or with 'encrypt' and our own message.
+-- >   -- Do not reuse a nonce with the same secret key!
+-- >   (nonce, encryptedMessage) <- AuthenticatedEncryption.encrypt "hello hello" secretKey
+-- >   let result = AuthenticatedEncryption.decrypt encryptedMessage secretKey nonce
+-- >   print result
+-- >   -- "Just \"hello hello\""
+
+-- | A secret key of size 'cryptoSecretboxKeyBytes'.
+--
+-- @since 0.0.1.0
+newtype SecretKey = SecretKey (ForeignPtr CUChar)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq SecretKey where
+  (SecretKey hk1) == (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretboxKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord SecretKey where
+  compare (SecretKey hk1) (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretboxKeyBytes
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Display SecretKey where
+  displayBuilder = Builder.fromText . secretKeyToHexText
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Show SecretKey where
+  show = BS.unpackChars . secretKeyToHexByteString
```

Comment:
Might be worth explaining why this is somewhat dangerous. I would actually argue that it might be better to have `Show` and `Display` both do censoring, and provide something like `unsafeDisplayNoDon'tDoThisWeMeanIt` to show it in hex.

Reply:
ah yes thank you